File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/J._Общий_подмассив/solution.py
import sys

MOD1: int = 100000003
MOD2: int = 100000019
MOD3: int = 100000037
q1: int = 100000009
q2: int = 100000007
q3: int = 100000021


def has_common_substring_with_length_k(k: int, A: list[int], B: list[int]) -> bool:
    if k == 0:
        return True
    if len(A) < k or len(B) < k:
        return False

    pool: set[tuple[int, int, int]] = set()
    cur_hash1: int = 0
    cur_hash2: int = 0
    cur_hash3: int = 0
    q_pow_k1: int = pow(q1, k, MOD1)
    q_pow_k2: int = pow(q2, k, MOD2)
    q_pow_k3: int = pow(q3, k, MOD3)

    for idx in range(len(A)):
        cur_hash1 = (cur_hash1 * q1 + A[idx]) % MOD1
        cur_hash2 = (cur_hash2 * q2 + A[idx]) % MOD2
        cur_hash3 = (cur_hash3 * q3 + A[idx]) % MOD3

        if idx >= k - 1:
            if idx >= k:
                cur_hash1 = (cur_hash1 - A[idx - k] * q_pow_k1) % MOD1
                cur_hash2 = (cur_hash2 - A[idx - k] * q_pow_k2) % MOD2
                cur_hash3 = (cur_hash3 - A[idx - k] * q_pow_k3) % MOD3

            pool.add((cur_hash1, cur_hash2, cur_hash3))

    cur_hash1 = 0
    cur_hash2 = 0
    cur_hash3 = 0
    for idx in range(len(B)):
        cur_hash1 = (cur_hash1 * q1 + B[idx]) % MOD1
        cur_hash2 = (cur_hash2 * q2 + B[idx]) % MOD2
        cur_hash3 = (cur_hash3 * q3 + B[idx]) % MOD3

        if idx >= k - 1:
            if idx >= k:
                cur_hash1 = (cur_hash1 - B[idx - k] * q_pow_k1) % MOD1
                cur_hash2 = (cur_hash2 - B[idx - k] * q_pow_k2) % MOD2
                cur_hash3 = (cur_hash3 - B[idx - k] * q_pow_k3) % MOD3

            if (cur_hash1, cur_hash2, cur_hash3) in pool:
                return True

    return False


def solution() -> None:
    """
    Time Complexity O( (NM)log(min(N, M) )
    Memory Complexity O(max(N, M))
    """
    n: int = int(sys.stdin.readline().rstrip())
    A: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    m: int = int(sys.stdin.readline().rstrip())
    B: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))

    left: int = 0
    right: int = min(n, m)
    max_common: int = 0
    while left <= right:
        middle: int = left + ((right - left) >> 1)
        if has_common_substring_with_length_k(middle, A, B):
            max_common = middle
            left = middle + 1
        else:
            right = middle - 1
    sys.stdout.write(str(max_common) + '\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/J._Общий_подмассив/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <tuple>
#include <functional>

using namespace std;

const long long MOD1 = 100000003;
const long long MOD2 = 100000019;
const long long MOD3 = 100000037;
const long long q1 = 100000009;
const long long q2 = 100000007;
const long long q3 = 100000021;

struct TupleHash {
    template <class T>
    size_t operator()(const tuple<T, T, T>& t) const {
        auto hash1 = hash<T>()(get<0>(t));
        auto hash2 = hash<T>()(get<1>(t));
        auto hash3 = hash<T>()(get<2>(t));
        return hash1 ^ (hash2 << 1) ^ (hash3 << 2);
    }
};

bool has_common_substring_with_length_k(int k, const vector<int>& A, const vector<int>& B) {
    if (k == 0) return true;
    if (A.size() < k || B.size() < k) return false;

    unordered_set<tuple<long long, long long, long long>, TupleHash> pool;
    long long cur_hash1 = 0, cur_hash2 = 0, cur_hash3 = 0;
    long long q_pow_k1 = 1, q_pow_k2 = 1, q_pow_k3 = 1;

    for (int i = 0; i < k; ++i) {
        q_pow_k1 = (q_pow_k1 * q1) % MOD1;
        q_pow_k2 = (q_pow_k2 * q2) % MOD2;
        q_pow_k3 = (q_pow_k3 * q3) % MOD3;
    }

    for (int idx = 0; idx < A.size(); ++idx) {
        cur_hash1 = (cur_hash1 * q1 + A[idx] + MOD1) % MOD1;
        cur_hash2 = (cur_hash2 * q2 + A[idx] + MOD2) % MOD2;
        cur_hash3 = (cur_hash3 * q3 + A[idx] + MOD3) % MOD3;

        if (idx >= k - 1) {
            if (idx >= k) {
                cur_hash1 = (cur_hash1 - A[idx - k] * q_pow_k1 + MOD1) % MOD1;
                cur_hash2 = (cur_hash2 - A[idx - k] * q_pow_k2 + MOD2) % MOD2;
                cur_hash3 = (cur_hash3 - A[idx - k] * q_pow_k3 + MOD3) % MOD3;
            }
            pool.insert({cur_hash1, cur_hash2, cur_hash3});
        }
    }

    cur_hash1 = 0;
    cur_hash2 = 0;
    cur_hash3 = 0;
    for (int idx = 0; idx < B.size(); ++idx) {
        cur_hash1 = (cur_hash1 * q1 + B[idx] + MOD1) % MOD1;
        cur_hash2 = (cur_hash2 * q2 + B[idx] + MOD2) % MOD2;
        cur_hash3 = (cur_hash3 * q3 + B[idx] + MOD3) % MOD3;

        if (idx >= k - 1) {
            if (idx >= k) {
                cur_hash1 = (cur_hash1 - B[idx - k] * q_pow_k1 + MOD1) % MOD1;
                cur_hash2 = (cur_hash2 - B[idx - k] * q_pow_k2 + MOD2) % MOD2;
                cur_hash3 = (cur_hash3 - B[idx - k] * q_pow_k3 + MOD3) % MOD3;
            }
            if (pool.count({cur_hash1, cur_hash2, cur_hash3})) {
                return true;
            }
        }
    }

    return false;
}

void solution() {
    /*
    Time Complexity O( (NM)log(min(N, m)) )
    Memory Complexity O(max(N, M))
    */
    int n;
    cin >> n;
    vector<int> A(n);
    for (int i = 0; i < n; i++) {
        cin >> A[i];
    }
    int m;
    cin >> m;
    vector<int> B(m);
    for (int i = 0; i < m; i++) {
        cin >> B[i];
    }

    int left = 0, right = min(n, m);
    int max_common = 0;
    while (left <= right) {
        int middle = left + (right - left) / 2;
        if (has_common_substring_with_length_k(middle, A, B)) {
            max_common = middle;
            left = middle + 1;
        } else {
            right = middle - 1;
        }
    }
    cout << max_common << endl;
}

int main() {
    solution();
    return 0;
}
